sdk/Pulumi/Core/Urn.cs

```diff
@@ -1,15 +1,83 @@
 // Copyright 2016-2019, Pulumi Corporation
 
 using System;
+using System.Diagnostics.CodeAnalysis;
 
 namespace Pulumi
 {
     /// <summary>
     /// An automatically generated logical URN, used to stably identify resources. These are created
     /// automatically by Pulumi to identify resources.  They cannot be manually constructed.
     /// </summary>
-    public static class Urn
+    public readonly struct Urn : IEquatable<Urn>
     {
+        readonly string _value;
+
+        public Urn(string value)
+        {
+            if (string.IsNullOrEmpty(value))
+            {
+                throw new ArgumentException("URN cannot be null or empty.", nameof(value));
+            }
+
+            _value = value;
+        }
+
+        /// <summary>
+        /// Implicitly converts a URN to a string.
+        /// </summary>
+        public static implicit operator string(Urn value) => value._value;
+
+        /// <summary>
+        /// Returns the string representation of this URN.
+        /// </summary>
+        override public string ToString() => _value;
+
+        /// <summary>
+        /// Determines whether the specified object is equal to the current URN.
+        /// </summary>
+        public override bool Equals([NotNullWhen(true)] object? obj)
+        {
+            if (obj is Urn other)
+            {
+                return Equals(this, other);
+            }
+            return false;
+        }
```

Comment:
`Equals(this, other)` is not the `Equals` you want; you want `Equals(other)`. These lines could be simplified as:

```suggestion
            => obj is Urn other && Equals(other);
```